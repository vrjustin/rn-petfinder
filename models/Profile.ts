export interface Profile {
  shouldOnboard: boolean;
  isRehydrated: boolean;
  userName?: string;
}

export default Profile;
