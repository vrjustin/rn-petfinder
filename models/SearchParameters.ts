interface SearchParameters {
  location: {zipCode: string};
  distance: number;
  tagsPreffered: string[];
}

export default SearchParameters;
