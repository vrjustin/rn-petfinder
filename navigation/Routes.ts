export const Routes = {
  PetTypes: 'PetTypes',
  Breeds: 'Breeds',
  Animals: 'Animals',
  Options: 'Options',
  AnimalDetails: 'AnimalDetails',
  Contact: 'Contact',
  Organizations: 'Organizations',
};
