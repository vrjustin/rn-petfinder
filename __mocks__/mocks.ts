import Animal from '../models/Animal';
import Breed from '../models/Breed';
import {PetType} from '../models/PetType';
import {AnimalResults} from '../services/apiService';

export const selectedAnimalMock: Animal = {
  id: 1,
  organization_id: 'org1',
  url: 'https://example.com',
  age: 'Young',
  gender: 'Female',
  size: 'Medium',
  coat: 'Short',
  name: 'Fluffy',
  description: 'A fluffy dog',
  status: 'Available',
  attributes: {
    declawed: null,
    house_trained: true,
    shots_current: true,
    spayed_neutered: false,
    special_needs: false,
  },
  breeds: {
    mixed: false,
    primary: 'Poodle',
    secondary: null,
    unknown: false,
  },
  colors: {
    primary: 'White',
    secondary: null,
    tertiary: null,
  },
  contact: {
    address: {
      address1: '123 Main St',
      address2: null,
      city: 'Anytown',
      country: 'USA',
      postcode: '12345',
      state: 'NY',
    },
    email: 'email@example.com',
    phone: '555-123-4567',
  },
  photos: [
    {
      full: 'https://example.com/full.jpg',
      large: 'https://example.com/large.jpg',
      medium: 'https://example.com/medium.jpg',
      small: 'https://example.com/small.jpg',
    },
  ],
  primary_photo_cropped: {
    full: 'https://example.com/full.jpg',
    large: 'https://example.com/large.jpg',
    medium: 'https://example.com/medium.jpg',
    small: 'https://example.com/small.jpg',
  },
  species: 'Dog',
  tags: ['Friendly', 'Playful'],
  isFavorite: false,
};

export const mockAnimalResults: AnimalResults = {
  animalsData: [selectedAnimalMock],
  pagination: {current_page: 1, total_pages: 1},
};

export const mockPetTypeDog: PetType = {
  displayName: 'Dog',
  name: 'Dog',
  coats: [],
  colors: [],
  genders: [],
  _links: {
    self: {href: ''},
    breeds: {href: ''},
  },
};

export const mockBreeds: Breed[] = [
  {
    name: 'Poodle',
    _links: {
      type: {href: ''},
    },
  },
  {
    name: 'German Shephard Dog',
    _links: {
      type: {href: ''},
    },
  },
];
