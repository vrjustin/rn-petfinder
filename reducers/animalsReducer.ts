import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import Animal from '../models/Animal';

interface AnimalsState {
  animals: Animal[];
  favorites: Animal[];
}

const initialState: AnimalsState = {
  animals: [],
  favorites: [],
};

const animalsSlice = createSlice({
  name: 'animals',
  initialState,
  reducers: {
    setAnimals(state, action: PayloadAction<Animal[]>) {
      state.animals = action.payload;
    },
    toggleFavorite(state, action: PayloadAction<Animal>) {
      const isFavorite = state.favorites.some(
        fav => fav.id === action.payload.id,
      );
      if (isFavorite) {
        state.favorites = state.favorites.filter(
          fav => fav.id !== action.payload.id,
        );
      } else {
        state.favorites.push(action.payload);
      }
    },
  },
});

export const selectAnimals = (state: RootState) => state.animals.animals;
export const selectFavorites = (state: RootState) => state.animals.favorites;

export const {setAnimals, toggleFavorite} = animalsSlice.actions;
export default animalsSlice.reducer;
