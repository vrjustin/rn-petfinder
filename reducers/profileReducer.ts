import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import Profile from '../models/Profile';

interface ProfileState {
  profile: Profile;
}

const initialState: ProfileState = {
  profile: {
    shouldOnboard: true,
    isRehydrated: false,
  },
};

const profileSlice = createSlice({
  name: 'profile',
  initialState,
  reducers: {
    setProfile(state, action: PayloadAction<Profile>) {
      state.profile = action.payload;
    },
  },
});

export const profile = (state: RootState) => state.profile.profile;

export const {setProfile} = profileSlice.actions;
export default profileSlice.reducer;
