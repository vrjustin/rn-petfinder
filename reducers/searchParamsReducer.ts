import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import SearchParameters from '../models/SearchParameters';

interface SearchParametersState {
  searchParameters: SearchParameters;
}

const initialState: SearchParametersState = {
  searchParameters: {
    location: {
      zipCode: '90210',
    },
    distance: 500, //in miles
    tagsPreferred: [],
  },
};

const searchParametersSlice = createSlice({
  name: 'searchParameters',
  initialState,
  reducers: {
    setSearchParameters(state, action: PayloadAction<SearchParameters>) {
      state.searchParameters = action.payload;
    },
  },
});

export const selectSearchParameters = (state: RootState) =>
  state.searchParameters.searchParameters;

export const {setSearchParameters} = searchParametersSlice.actions;
export default searchParametersSlice.reducer;
