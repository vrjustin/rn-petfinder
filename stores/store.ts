import {combineReducers, configureStore} from '@reduxjs/toolkit';
import petTypesReducer from '../reducers/petTypesReducer';
import petBreedsReducer from '../reducers/petBreedsReducer';
import animalsReducer from '../reducers/animalsReducer';
import searchParamsReducer from '../reducers/searchParamsReducer';

const createDebugger = require('redux-flipper').default;

const rootReducer = combineReducers({
  petTypes: petTypesReducer,
  petBreeds: petBreedsReducer,
  animals: animalsReducer,
  searchParameters: searchParamsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware =>
    __DEV__
      ? getDefaultMiddleware({serializableCheck: false}).concat(
          createDebugger(),
        )
      : getDefaultMiddleware({serializableCheck: false}),
});

export default store;
