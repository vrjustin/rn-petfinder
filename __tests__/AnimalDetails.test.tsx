import React from 'react';
import {Provider} from 'react-redux';
import * as reactRedux from 'react-redux';
import store from '../stores/store';
import renderer, {act} from 'react-test-renderer';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import AnimalDetails from '../Views/AnimalDetails';
import {Routes} from '../navigation/Routes';
import {selectedAnimalMock} from '../__mocks__/mocks';
import SearchParameters from '../models/SearchParameters';
import * as searchParamsReducer from '../reducers/searchParamsReducer';
import * as profileReducer from '../reducers/profileReducer';
import * as animalsReducer from '../reducers/animalsReducer';
import Profile from '../models/Profile';
import AsyncStorage from '@react-native-async-storage/async-storage';

const mockNavigate = jest.fn();
const mockSetOptions = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: mockNavigate,
    setOptions: mockSetOptions,
  }),
}));

const mockSearchParameters: SearchParameters = {
  location: {
    zipCode: '90210',
  },
  distance: 5,
  tagsPreferred: ['Friendly'],
  breedsPreferred: [],
  orgsPagination: {
    currentPage: 1,
    totalPages: 1,
  },
  animalsPagination: {
    currentPage: 1,
    totalPages: 1,
  },
};

const mockProfile: Profile = {
  shouldOnboard: false,
  isRehydrated: true,
  userName: 'Test-User',
  signInMethod: 'google',
};

let renderedAnimalDetailsTree: any;

describe('AnimalDetails', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', async () => {
    await act(async () => {
      renderedAnimalDetailsTree = renderer.create(
        <Provider store={store}>
          <AnimalDetails
            route={{
              key: 'mockKey',
              name: 'Animal',
              params: {
                selectedAnimal: selectedAnimalMock,
              },
            }}
          />
        </Provider>,
      );
    });
    expect(renderedAnimalDetailsTree.toJSON()).toMatchSnapshot();
  });

  it('navigates to contact screen on press of contact details button', async () => {
    render(
      <Provider store={store}>
        <AnimalDetails
          route={{
            key: 'mockKey',
            name: 'Animal',
            params: {
              selectedAnimal: selectedAnimalMock,
            },
          }}
        />
      </Provider>,
    );

    await waitFor(() => expect(mockSetOptions).toHaveBeenCalled());

    //This will manually call the render function for the headerRight
    //of react navigation. this was the only way to get this to work.
    const headerRight = mockSetOptions.mock.calls[0][0].headerRight;
    const {getByTestId: getByTestIdHeader} = render(headerRight());

    const contactDetailsButton = getByTestIdHeader(
      'AnimalDetails-ContactOptionsButton',
    );
    fireEvent.press(contactDetailsButton);
    expect(mockNavigate).toHaveBeenCalledWith(Routes.Contact, {
      selectedAnimal: selectedAnimalMock,
    });
  });

  it('adds the tag to the tagsPreferred searchParameters', async () => {
    const mockDispatch = jest.fn();
    jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(mockDispatch);

    const {getByTestId} = render(
      <Provider store={store}>
        <AnimalDetails
          route={{
            key: 'mockKey',
            name: 'Animal',
            params: {
              selectedAnimal: selectedAnimalMock,
            },
          }}
        />
      </Provider>,
    );

    const friendlyTag = getByTestId(`AnimalDetailsTag-Friendly`);

    fireEvent.press(friendlyTag);

    const dispatchedAction = mockDispatch.mock.calls[0][0];
    expect(dispatchedAction.type).toBe('searchParameters/setSearchParameters');
    expect(dispatchedAction.payload.tagsPreferred).toEqual([
      selectedAnimalMock.tags[0],
    ]);
  });

  it('removes the tag from the tagsPreferred searchParameters', async () => {
    const mockDispatch = jest.fn();
    jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(mockDispatch);
    jest
      .spyOn(searchParamsReducer, 'selectSearchParameters')
      .mockReturnValue(mockSearchParameters);

    const {getByTestId} = render(
      <Provider store={store}>
        <AnimalDetails
          route={{
            key: 'mockKey',
            name: 'Animal',
            params: {
              selectedAnimal: selectedAnimalMock,
            },
          }}
        />
      </Provider>,
    );

    const friendlyTag = getByTestId(`AnimalDetailsTag-Friendly`);

    fireEvent.press(friendlyTag);

    const dispatchedAction = mockDispatch.mock.calls[0][0];
    expect(dispatchedAction.type).toBe('searchParameters/setSearchParameters');
    expect(dispatchedAction.payload.tagsPreferred).toEqual([]);
  });

  it('handles favorite properly for selectedAnimal', async () => {
    const mockDispatch = jest.fn();
    jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(mockDispatch);
    jest.spyOn(profileReducer, 'profile').mockReturnValue(mockProfile);
    jest
      .spyOn(animalsReducer, 'selectAnimals')
      .mockReturnValue([selectedAnimalMock]);

    const {getByTestId} = render(
      <Provider store={store}>
        <AnimalDetails
          route={{
            key: 'mockKey',
            name: 'Animal',
            params: {
              selectedAnimal: selectedAnimalMock,
            },
          }}
        />
      </Provider>,
    );
    const favoriteButton = getByTestId(`AnimalDetails-FavoriteButton`);
    fireEvent.press(favoriteButton);

    const dispatchedAction = mockDispatch.mock.calls[0][0];
    expect(dispatchedAction.type).toBe('animals/setAnimals');
    const favoritedAnimalAfterDispatch = dispatchedAction.payload[0];
    expect(favoritedAnimalAfterDispatch.isFavorite).toBe(true);
    expect(AsyncStorage.setItem).toHaveBeenCalledWith(
      'animals',
      JSON.stringify(dispatchedAction.payload),
    );
  });

  it('sets the currentImageIndex onPress of gallery item', async () => {
    const {getByTestId} = render(
      <Provider store={store}>
        <AnimalDetails
          route={{
            key: 'mockKey',
            name: 'Animal',
            params: {
              selectedAnimal: selectedAnimalMock,
            },
          }}
        />
      </Provider>,
    );
    const galleryItem1 = getByTestId(`AnimalDetailsGalleryItem-1`);
    const startingStyle = galleryItem1.props.children[0].props.style;
    expect(startingStyle).toStrictEqual({width: 100, height: 100, margin: 5});
    fireEvent.press(galleryItem1);
    const selectedStyle = galleryItem1.props.children[0].props.style;
    expect(selectedStyle).toStrictEqual({
      width: 100,
      height: 100,
      margin: 5,
      borderWidth: 2,
      borderColor: 'red',
    });
  });
});
