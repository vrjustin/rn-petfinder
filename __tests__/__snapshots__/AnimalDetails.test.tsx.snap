// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimalDetails renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "height": "50%",
        "overflow": "hidden",
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        {
          "flex": 1,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          {
            "uri": "https://example.com/large.jpg",
          }
        }
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "height": undefined,
              "width": undefined,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          {
            "backgroundColor": "gray",
            "borderRadius": 8,
            "bottom": 8,
            "left": 8,
            "padding": 4,
            "position": "absolute",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "flex-end",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 24,
                "fontWeight": "bold",
              }
            }
          >
            Fluffy
            , 
          </Text>
          <Text
            style={
              {
                "color": "white",
                "fontSize": 16,
              }
            }
          >
            Young
          </Text>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "white",
                  "fontSize": 20,
                },
                {
                  "marginRight": 8,
                },
                {
                  "fontFamily": "FontAwesome",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            ï‚¬
          </Text>
          <Text
            style={
              {
                "color": "white",
                "fontSize": 16,
              }
            }
          >
            Anytown
            , 
            NY
          </Text>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView>
    <View>
      <Text
        style={
          {
            "fontSize": 16,
            "fontWeight": "bold",
            "padding": 8,
          }
        }
      >
        Gallery
      </Text>
      <RCTScrollView
        data={
          [
            {
              "full": "https://example.com/full.jpg",
              "large": "https://example.com/large.jpg",
              "medium": "https://example.com/medium.jpg",
              "small": "https://example.com/small.jpg",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={true}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              [
                {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    {
                      "uri": "https://example.com/small.jpg",
                    }
                  }
                  style={
                    {
                      "borderColor": "red",
                      "borderWidth": 2,
                      "height": 100,
                      "margin": 5,
                      "width": 100,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <Text
        style={
          {
            "fontSize": 16,
            "fontWeight": "bold",
            "padding": 8,
          }
        }
      >
        Introduction
      </Text>
      <Text
        style={
          {
            "fontSize": 14,
            "padding": 8,
          }
        }
      >
        A fluffy dog
      </Text>
      <View>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "bold",
              "padding": 8,
            }
          }
        >
          Attributes
        </Text>
        <View
          style={
            {
              "padding": 16,
              "paddingRight": 32,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "#f0f0f0",
                "borderRadius": 8,
                "padding": 8,
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingVertical": 4,
                }
              }
            >
              <Text
                style={
                  {
                    "fontWeight": "normal",
                  }
                }
              >
                House Trained
              </Text>
              <Text
                style={{}}
              >
                Yes
              </Text>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingVertical": 4,
                }
              }
            >
              <Text
                style={
                  {
                    "fontWeight": "normal",
                  }
                }
              >
                Shots Current
              </Text>
              <Text
                style={{}}
              >
                Yes
              </Text>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingVertical": 4,
                }
              }
            >
              <Text
                style={
                  {
                    "fontWeight": "normal",
                  }
                }
              >
                Spayed/Neutered
              </Text>
              <Text
                style={{}}
              >
                No
              </Text>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingVertical": 4,
                }
              }
            >
              <Text
                style={
                  {
                    "fontWeight": "normal",
                  }
                }
              >
                Special Needs
              </Text>
              <Text
                style={{}}
              >
                No
              </Text>
            </View>
          </View>
        </View>
      </View>
      <Text
        style={
          {
            "fontSize": 16,
            "fontWeight": "bold",
            "padding": 8,
          }
        }
      >
        Tags
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "padding": 8,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "gray",
                "borderRadius": 8,
                "marginBottom": 8,
                "marginRight": 8,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              }
            }
          >
            <Text
              style={
                {
                  "color": "white",
                }
              }
            >
              Friendly
            </Text>
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "gray",
                "borderRadius": 8,
                "marginBottom": 8,
                "marginRight": 8,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              }
            }
          >
            <Text
              style={
                {
                  "color": "white",
                }
              }
            >
              Playful
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
